src/Uno.UI/UI/Xaml/ResourceResolver.cs

```diff
@@ -0,0 +1,218 @@
+		/// <returns>True if the resource was found, false if not.</returns>
+		private static bool TryTopLevelRetrieval(object resourceKey, out object value)
+		{
+			value = null;
+			return (Application.Current?.Resources.TryGetValue(resourceKey, out value) ?? false) ||
+				MasterDictionary.TryGetValue(resourceKey, out value);
+		}
+
+		/// <summary>
+		/// Get a system-level resource with the given key.
+		/// </summary>
+		internal static T GetSystemResource<T>(object key)
+		{
+			if (MasterDictionary.TryGetValue(key, out var value) && value is T t)
+			{
+				return t;
+			}
+
+			return default(T);
+		}
+
+		/// <summary>
+		/// Push a new <see cref="XamlScope"/>, typically because a template is being materialized.
+		/// </summary>
+		/// <param name="scope"></param>
+		internal static void PushNewScope(XamlScope scope) => _scopeStack.Push(scope);
+		/// <summary>
+		/// Push a new Resources source to the current xaml scope.
+		/// </summary>
+		internal static void PushSourceToScope(DependencyObject source) => PushSourceToScope((source as IWeakReferenceProvider).WeakReference);
+		/// <summary>
+		/// Push a new Resources source to the current xaml scope.
+		/// </summary>
+		internal static void PushSourceToScope(ManagedWeakReference source)
+		{
+			var current = _scopeStack.Pop();
+			_scopeStack.Push(current.Push(source));
+		}
+		/// <summary>
+		/// Pop Resources source from current xaml scope.
+		/// </summary>
+		internal static void PopSourceFromScope()
+		{
+			var current = _scopeStack.Pop();
+			_scopeStack.Push(current.Pop());
+		}
+		/// <summary>
+		/// Pop current <see cref="XamlScope"/>, typically because template materialization is complete.
+		/// </summary>
+		internal static void PopScope()
+		{
+			_scopeStack.Pop();
+			if (_scopeStack.Count == 0)
+			{
+				throw new InvalidOperationException("Base scope should never be popped.");
+			}
+		}
+
+		/// <summary>
+		/// If tracing is enabled, writes an event for the initialization of system-level resources (Generic.xaml etc)
+		/// </summary>
+		internal static IDisposable WriteInitiateGlobalStaticResourcesEventActivity() => _trace.WriteEventActivity(TraceProvider.InitGenericXamlStart, TraceProvider.InitGenericXamlStop);
+
+		/// <summary>
+		/// Register a dictionary for a given source, this is used for retrieval when setting the Source property in code-behind or to an
+		/// external resource.
+		/// </summary>
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public static void RegisterResourceDictionaryBySource(string uri, Func<ResourceDictionary> dictionary)
+		{
+			_registeredDictionaries[uri] = dictionary;
+		}
+
+		/// <summary>
+		/// Retrieve the ResourceDictionary mapping to a given source. Throws an exception if none is found.
+		/// </summary>
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public static ResourceDictionary RetrieveDictionaryForSource(Uri source)
+		{
+			if (_registeredDictionaries.TryGetValue(source.AbsoluteUri, out var factory))
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Refactor this method to add validation of parameter 'source' before using it.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=4826539)